### ISSUE TITLE
A Tail-Index Analysis of Stochastic Gradient Noise in Deep Neural Networks

### ISSUE THREAD

**Issue by nohzen** [nohzen/paper_notes#1]
# 概要
SGDが、Gradient Descentより汎化性能が高くなる理由を説明した。

# 論文情報
- ICML2019 [Honorable Mentions](https://medium.com/syncedreview/icml-2019-google-eth-zurich-mpi-is-cambridge-prowler-io-share-best-paper-honours-4aeabd5c9fc8)受賞
- 著者: Umut Simsekli (Paris-Saclay) _et al._
- paper: 
http://proceedings.mlr.press/v97/simsekli19a.html
https://arxiv.org/abs/1901.06053
- code: https://github.com/umutsimsekli/sgd_tail_index

# 内容
## 背景
- SGDは元々、学習データが多いと一度にLossの勾配を計算できないという数値的な理由で使われ始めた。
- SGDはGradient Descentに比べ、**汎化性能(テストデータでの精度)が高くなる**ことが**実験的に**知られている[Keskar+(2016)]
- SGDを使うと汎化性能が上がる理由の説明を試みた先行研究[Jastrzebski+(2017)]は存在するが、ノイズの性質が実験と整合しないなど、問題点があった。

## 理論
- 学習Lossで、フラットなminimumの方が、シャープなminimumより汎化性能が高い
  - 直感的には、以下の図のように、学習LossとテストLossのミスマッチに対してロバストになるため
  - そもそも、Lossのフラットさの定義や、汎化性能との関係についてはコンセンサスが得られている訳では無さそう。Ref: https://www.inference.vc/sharp-vs-flat-minima-are-still-a-mystery-to-me/
  - 本論文では、フラットなminimumの方が汎化性能が高くなることは仮定する。したがって、SGDを使うとフラットなminimumに行きやすいということが示せれば、「SGDを使うと汎化性能が上がる」が示せたことになる。
![flat_minimum](https://user-images.githubusercontent.com/29157627/68147325-29df3680-ff7d-11e9-8b9e-7b72b663f40a.png) 図は[Keskar+(2016)]より
- **中心極限定理**が使えるとすると、バッチサイズが十分大きい時には、SGDはGradient Descentに正規分布に従うノイズが加わったものとみなせる。さらに、学習率をゼロにする極限で、SGDは確率微分方程式(SDE, ブラウン運動)になる。
  - この場合、[Jastrzebski+(2017)]で示されているように、汎化性能は バッチサイズ/学習率 が小さいほど高くなる。
  - [Jastrzebski+(2017)]の議論は、以下のような問題がある。
    - minimumからの脱出時間は、minimumの幅のべき、深さの指数関数になっている。つまり、深いminimum(シャープなminimum)に行ってしまう。
    - SGDのノイズが実際の実験では正規分布に見えない
![noise](https://user-images.githubusercontent.com/29157627/68149191-c8b96200-ff80-11e9-8b39-48589531450a.png)
    - 議論が成り立つには、ネットワークのパラメータ数の指数関数のiterationが必要
- 中心極限定理は、確率分布(今の場合は、学習データを生成する分布)の分散が有限の時にしか適用できない。そこで、分散が発散している場合にも適用可能な**一般化中心極限定理**を使う。
  - SGDはGradient Descentに、**α-安定分布**(裾がべきα+1で減衰する分布)に従うノイズが加わったものとみなせる。学習率ゼロの極限で、SGDはα-安定分布のノイズのある確率微分方程式(**Revy運動**)になる。
  - 正規分布に比べ、α-安定分布は大きな値を取りやすく、"jump"と呼ばれる急激な変化が見られる(深さに関わらず、minimumから一気に脱出できる！)
  - 簡単化した設定では、minimunからの脱出時間は、munimumの幅とαのべきになる。
## 実験
- データセット: MNIST, CIFAR10, CIFAR100
- ネットワーク: fully connected network(FCN), CNN(AlexNet)
1. αの見積もり
α-安定分布はtail-index αが、2の時は正規分布、2より小さい時は裾がべき減衰する。
数値実験により、αはデータセット/ネットワークの種類/ネットワークのサイズによらず、2より小さいことが確認できた。
<img src="https://user-images.githubusercontent.com/29157627/68150685-9a895180-ff83-11e9-99f2-e6123ff9e3de.png" width="400">
2. バッチサイズの影響
バッチサイズを大きくしても、αは変化しない。
つまり、バッチサイズが小さいために、中心極限定理が成り立っていないのではない。（元の確率分布の分散が発散しているために中心極限定理がなりなっていない。）
<img src="https://user-images.githubusercontent.com/29157627/68150948-1daaa780-ff84-11e9-8203-8f6c9ce71a45.png" width="400">
3. jump
↑の実験のαは、最終的な値。学習途中でのαは、以下のようになる。
Lossの丘を越えているために特異的な振る舞いが見られる？
<img src="https://user-images.githubusercontent.com/29157627/68151071-61051600-ff84-11e9-81e6-99355aa8ec0d.png" width="600">

# コメント/感想/批判
- この論文を踏まえて、汎化性能が高くなるようにSGDの改良版みたいなものを作れない？
- [Jastrzebski+(2017)]では汎化性能は バッチサイズ/学習率 の関数(minimumの幅と高さにも依存する)だった。この論文の結論では、汎化性能はαとminimumの幅のべきになる。汎化性能がバッチサイズや学習率に全く依存しないのは実験と矛盾するのでは？
- α-安定分布のノイズの確率微分方程式の解析が甘そう。α固定にしてたり。

# 引用/関連研究
- [N. S. Keskar *et al.*, "On Large-Batch Training for Deep Learning: Generalization Gap and Sharp Minima" (2016).](https://arxiv.org/abs/1609.04836) SGDだと汎化性能があがる
- [S. Hochreiter & J. Schmidhuber, "Flat minima" (1997).](https://www.bioinf.jku.at/publications/older/3304.pdf) フラットな底は、シャープな底より汎化性能が高い
- [S. Jastrzebski *et al.*, "Three Factors Influencing Minima in SGD" (2017)](https://arxiv.org/abs/1711.04623) SGDを使うと汎化性能が上がる理由の説明(正規分布のノイズを仮定)
